### DISCUSSION TITLE
Understanding Memory Buffers and ``donate_argnums``

### DISCUSSION THREAD

**Discussion by Jeevesh8** [jax-ml/jax#7122] [Q&A]
I have a setup of the form:
```python
@partial(jax.pmap, donate_argnums=(0))
def fn(x):
    new_x = 2+x
    return new_x, jnp.mean(x)

@jax.pmap
def fn2(x):
    return x, jnp.mean(x)

x = jnp.array([[1,2,3,4,5,6,7,8]*800000]*8)

for i in range(1,4):
    print(x)
    x, y = fn(x)
    if i%2==0:
        for j in range(3):
            print(x)
            x, y = fn2(x)
```
Now, my problem is, when we enter the second loop, I end up getting a ``RuntimeError: Invalid argument: DeviceArray has been deleted.``,  when the first call to ``fn2`` occurs. I am able to print ``x`` just before entering ``fn2``. That means, that ``x`` has not been deleted, but still it says that it has been deleted. I want to know why?

**Note:** The above example doesn't reproduce the error, and is just representative of the flow of actual computation I am doing. 

Any suggestions as to how to debug this are also welcome.

**Comment by jakevdp** [comment]
There is some information on `donate_argnums` in the pmap [docstring](https://jax.readthedocs.io/en/latest/jax.html). To quote:

> Specify which arguments are “donated” to the computation. It is safe to donate arguments if you no longer need them once the computation has finished. In some cases XLA can make use of donated buffers to reduce the amount of memory needed to perform a computation, for example recycling one of your input buffers to store a result. You should not reuse buffers that you donate to a computation, JAX will raise an error if you try to. By default, no arguments are donated.

So to paraphrase, arguments marked for donation may be used by XLA to make the computation more efficient, in which case the buffers will be marked deleted. The reason you can print the contents of a deleted buffer is I think because of internal cacheing used for printing device arrays that ends up bypassing the deletion check; I would probably consider this a bug.

Edit: I just looked and the code path I suspected would lead to this behavior does not. Can you give an example of a short code snippet that reproduces the issue you're seeing? The one in your question does not seem to result in any error.

**Reply by Jeevesh8** [reply]
Sadly, I am unable to reproduce the error, in a short code snippet as of now. All the small examples I built, till now, end up running successfully. I even tried adding the flax parts of my code, but still I am unable to reproduce error. 
I am originally running on TPU VM, and trying to reproduce on colab.
I would probably have been able to reproduce.. had I known the source of it.
That is why I wanted to know, like, how much access from buffers is allowed? How are buffers working exactly? Is there some way to track where exactly the problem I am facing is coming from?

I feel bad saying this, but, if you'd want to look at my original code: the calls from [this](https://github.com/Jeevesh8/relational-pt/blob/91797b642ca2f60b7a44947aa0af81090bd1a03a/train_script.py#L311) to [this](https://github.com/Jeevesh8/relational-pt/blob/91797b642ca2f60b7a44947aa0af81090bd1a03a/train_script.py#L184) to [this](https://github.com/Jeevesh8/relational-pt/blob/91797b642ca2f60b7a44947aa0af81090bd1a03a/train_script.py#L133) leads to the error. The state is being propagated from [here](https://github.com/Jeevesh8/relational-pt/blob/91797b642ca2f60b7a44947aa0af81090bd1a03a/train_script.py#L284) into the ``eval_step``.

You may choose not to look.. I am sorry for this. Will let you know if I figure out where the error is coming from.

**Reply by jakevdp** [reply]
The general advice would be: don't use `donate_argnums` for buffers that you plan to use again. At best it has no effect, at worst it restuls in an error.

**Reply by Jeevesh8** [reply]
If I return the thing out of the function being ``pmap``-ed, and update its value in my loop.. like ``x`` above; I should be able to access the returned ``x`` any which way I want, right? Or is it considered part of buffer even after being returned? Like it can only go back into the function it came from.. or something like that?

**Reply by jakevdp** [reply]
Arguments marked donated can be used by the XLA compiler however it wants. As far as I understand it, XLA may even decide to use the input buffer as the output buffer in some cases. But you should not rely on that, and JAX's error in this case is there to prevent you from shooting yourself in the foot.

**Reply by Jeevesh8** [reply]
I think my problem is coming maybe because of asynchronized approach of Jax. That is, I take the output(say ``o``) from the first function, for example, after its first two runs, in the above code, and try to pass it to the function in the inner loop. But while the function in the inner loop is getting compiled, the first function starts running again, with its input ``x`` as ``o``, as a result ``o`` gets cleared out, before the second(inner) loop gets to access it.

Earlier I used to believe that ``print(x)`` would block till ``x`` becomes available, so I thought if my second function had ``print()`` like statements(after the first call to function being ``pmap``-ed), that the above situation won't happen. But, when I ran some code today, that logged the avg. loss of a model at regular intervals, while training.. and piped its output to a file; I observe that the file gets filled in "bursts", where 10-20 logs for loss comes at once, with long durations in between, during which nothing is logged. 

**EDIT:** The ``print()`` is blocking. But the write to file, is often buffered, which leads to data being logged in bursts.
**EDIT2:** Also, the compile of a function is blocking. So above argument is invalid.

**Reply by jakevdp** [reply]
If you mark an array as `donate_argnums`, you should not use that array any more. If you want to use the same array later, don't mark it for donation. One of those two approaches should solve your problem.

**Reply by Jeevesh8** [reply]
Ohh.. but I see that the ``async`` problem may even occur even if the second function has ``print`` or other blocking statements. For if the first function doesn't have any, it will go on executing in a non-blocking way, and when the control reaches for calling to the function being ``pmap``-ed(the "second function") for the first time, will the second function wait for the output from the first to become available ? And then continue, to compile? I think it will compile from just the tracer values(or whatever the first function outputs instead of the actual computed outputs) output from the first function, but just after it compiles, the state it will try to access, will it wait for it to become available?

**Reply by Jeevesh8** [reply]
@jakevdp I am not using the same array, but trying to use the array returned in place of it. But I see that that too gets donated in the next step.. so probably not safe to use it, unless I ``block_until_ready()`` and stop it from getting donated in the next iteration, and only after using it in the internal function, donate it for the next iteration.

**Reply by jakevdp** [reply]
Oh, I see – sorry I missed that aspect of it.

I think it would be worth taking a deeper look at what's going on here – if you're able to put together a minimal repro, it would be great if you could open an issue with the report.

**Reply by Jeevesh8** [reply]
How can I check if a device array has been deleted or not?

**Reply by jakevdp** [reply]
```python
import jax.numpy as jnp
x = jnp.arange(10)
print(x.is_deleted())
# False
```
